### PR TITLE
Make discovery independent of the azure dependencies

### PR DESCRIPTION
**Pull request by roidelapluie**
For #7107

@bwplotka what do you think of this approach?

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

File: cmd/prometheus/main.go [hunk 1]
```diff
@@ -52,6 +52,8 @@ import (
 	promlogflag "github.com/prometheus/common/promlog/flag"
 	"github.com/prometheus/prometheus/config"
 	"github.com/prometheus/prometheus/discovery"
+
+	_ "github.com/prometheus/prometheus/discovery/azure"
```
Comment (bwplotka) [inline cmd/prometheus/main.go:56]: ?

File: discovery/manager.go [hunk 2]
```diff
@@ -418,7 +443,11 @@ func (m *Manager) registerProviders(cfg sd_config.ServiceDiscoveryConfig, setNam
 	}
 	for _, c := range cfg.AzureSDConfigs {
 		add(c, func() (Discoverer, error) {
-			return azure.NewDiscovery(c, log.With(m.logger, "discovery", "azure")), nil
+			d, err := GetDiscoverer("azure")
```
Comment (bwplotka) [inline discovery/manager.go:446]: Can we adjust `GetDiscoverer` in a way that would not require this anonymous function? :thinking:

File: discovery/manager.go [hunk 3]
```diff
@@ -78,12 +77,38 @@ var (
...
+		return fmt.Errorf("provider %s is already registered", name)
+	}
+	providers[name] = d
+	return nil
+}
+
+// GetDiscoverer return a discoverer that was registered with
+// RegisterDiscoverer.
+func GetDiscoverer(name string) (func(interface{}, log.Logger) Discoverer, error) {
+	registerMtx.Lock()
```
Comment (bwplotka) [inline discovery/manager.go:103]: I don't think init is ever done concurrently so we don't need any mutex here and above

Comment (roidelapluie) [reply discovery/manager.go:103]: Do you think that all the init() functions are called in one go routing?

Comment (bwplotka) [reply discovery/manager.go:103]: yea I think so

**Comment by bwplotka** [comment]
So BTW it's all about what is the requirements / thing we want achieve.

I don't think we can avoiding importing all of those deps from `main`. However we want to remove spread from packages that can be importable. To achieve this I don't think we need dynamic import registering. I think there is a type safe method to do this :hugs:

**Comment by brian-brazil** [comment]
Anonymous imports like this are the standard Go way to do this, and we use it in other places.

**Pull request closed without merging**